### PR TITLE
fix: lock invocation per function for not having concurrent requests

### PR DESCRIPTION
**Pull request by mndeveci**
#### Which issue(s) does this change fix?
#6157

#### Why is this change necessary?
When running `sam local start-api` SAM CLI spins up one container per lambda function. Any concurrent request might fail (or have side-effects) due to this limitation.

#### How does it address the issue?
Create lock per function endpoint and only allow single execution per function.

#### What side effects does this change have?
N/A

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/local/docker/container.py [hunk 1]
```diff
@@ -378,11 +383,22 @@ def wait_for_http_response(self, name, event, stdout) -> Tuple[Union[str, bytes]
...
 
-        resp = requests.post(
-            self.URL.format(host=self._container_host, port=self.rapid_port_host, function_name="function"),
-            data=event.encode("utf-8"),
-            timeout=(self.RAPID_CONNECTION_TIMEOUT, None),
-        )
+        # generate a lock key with host-port combination which is unique per function
+        lock_key = f"{self._container_host}-{self.rapid_port_host}"
+        LOG.debug("Getting lock for the key %s", lock_key)
+        with CONCURRENT_CALL_MANAGER_LOCK:
```
Comment (mildaniel) [inline samcli/local/docker/container.py:389]: Why do we need this lock and not just the `CONCURRENT_CALL_MANAGER`?

File: tests/integration/local/start_api/test_start_api.py [hunk 2]
```diff
@@ -164,7 +165,7 @@ def setUp(self):
         self.url = "http://127.0.0.1:{}".format(self.port)
         HTTPConnection._http_vsn_str = "HTTP/1.1"
 
-    @pytest.mark.flaky(reruns=3)
+    # @pytest.mark.flaky(reruns=3)
```
Comment (mildaniel) [inline tests/integration/local/start_api/test_start_api.py:168]: Testing comment?

File: samcli/local/docker/container.py (hunk 1 above)
Comment (mndeveci) [reply samcli/local/docker/container.py:389]: We need locks for each lambda container, and in order to create/access that lock we need another one. So `CONCURRENT_CALL_MANAGER_LOCK` locks the first call to either create/access the lock for certain lambda function. And that accessed/created lock will actually lock the RIE call below.

File: tests/integration/local/start_api/test_start_api.py (hunk 2 above)
Comment (mndeveci) [reply tests/integration/local/start_api/test_start_api.py:168]:
```suggestion
    @pytest.mark.flaky(reruns=3)
```

**Review by mildaniel — APPROVED** [review]

**Review by hnnasit — APPROVED** [review]

**Pull request merged**